jsparagus/parse_table.py

```diff
@@ -1378,6 +1518,28 @@ def all_action_inedges(s: StateAndTransitions) -> bool:
         self.states.extend(from_act_action_states)
         self.rewrite_reordered_state_indexes()
 
+    def group_nonterminal_states(self, verbose: bool, progress: bool) -> None:
+        # This function is used to reduce the range of FilterStates values,
+        # such that the Rust compiler can compile FilterStates match statements
+        # to a table-switch.
+        freq_count = collections.Counter(nt for s in self.states for nt in s.nonterminals)
+        freq_nt, _ = zip(*freq_count.most_common())
+
+        def state_value(s: StateAndTransitions) -> float:
+            value = 0.0
+            i = 1.0
+            if len(s.epsilon) != 0:
+                return 2.0
+            if len(s.nonterminals) == 0:
+                return 1.0
+            for nt in freq_nt:
+                if nt in s:
+                    value += i
+                i /= 2.0
+            return -value
```

Comment:
Why the minus sign here? It seems OK to put the most common states up front, terminal-only states after that, and epsilon states at the end, but the minus sign reverses that, doesn' it?

Reply:
using a minus sign is used as a way to ensure that the most frequent non-terminals are listed with first in the list of states. Thus with smaller indexes.

Removing the minus sign would cause the most frequent non-terminals to be listed last among the list of states which have some non-terminals.

With negative sign:
```rust
pub fn full_actions_2192<'alloc, Handler>(parser: &mut Handler) -> Result<'alloc, bool>
where
    Handler: ParserTrait<'alloc, StackValue<'alloc>> + AstBuilderDelegate<'alloc>
{
        match parser.top_state() {
            160 | 161 | 162 | 163 | 165 | 166 | 167 | 168 | 170 | 171 | 172 | 173 | 174 | 175 | 176 | 177 | 178 | 179 | 180 | 181 | 182 | 183 | 184 | 185 | 187 | 188 | 189 | 190 | 191 | 192 | 193 | 194 | 195 | 196 | 197 | 199 | 200 | 201 | 202 | 203 | 204 | 205 | 206 | 207 | 208 | 209 | 210 | 211 | 212 | 213 | 214 | 215 | 216 | 217 | 218 | 219 | 220 | 221 | 222 | 223 | 224 | 225 | 226 | 227 | 228 => {
                // --> 2187
                full_actions_2187(parser)
            }
            80 | 81 | 82 | 83 | 84 | 85 | 86 | 87 | 88 | 89 | 90 | 91 | 92 | 93 | 94 | 95 | 96 | 97 | 98 | 99 | 100 | 101 | 102 | 103 | 104 | 105 | 106 | 107 | 108 | 109 | 110 | 111 | 112 | 113 | 114 | 115 | 116 | 117 | 118 | 119 | 120 | 121 | 122 | 123 | 124 | 125 | 126 | 127 | 128 | 129 | 130 | 131 | 132 | 133 | 134 | 135 | 136 | 137 | 138 | 139 | 140 | 141 | 142 | 143 | 144 | 145 | 146 | 147 | 148 | 149 | 150 | 151 | 152 | 153 | 154 | 155 | 156 => {
                // --> 2188
                full_actions_2188(parser)
            }
            230 | 231 | 232 | 233 | 234 | 235 | 236 | 237 | 238 | 239 | 240 | 241 | 242 | 243 | 244 | 245 | 246 | 247 | 248 | 249 | 250 | 251 | 252 | 253 | 254 | 255 | 256 | 257 | 258 | 259 | 260 | 261 | 262 | 263 | 264 | 265 | 266 | 267 | 268 | 269 | 270 | 271 | 272 | 273 | 274 | 275 | 276 | 277 | 278 | 279 | 280 => {
                // --> 2189
                full_actions_2189(parser)
            }
            0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 9 | 10 | 11 | 12 | 13 | 16 | 17 | 18 | 19 | 20 | 21 | 22 | 23 | 24 | 25 | 26 | 27 | 28 | 29 | 30 | 31 | 32 | 33 | 34 | 35 | 36 | 37 | 38 | 39 | 40 | 41 | 42 | 43 | 44 | 45 | 46 | 47 | 48 | 49 | 50 | 51 | 52 | 53 | 54 | 55 | 56 | 57 | 58 | 59 | 60 | 61 | 62 | 63 | 64 | 65 | 66 | 67 | 68 | 69 | 70 | 71 | 72 | 73 | 74 | 75 | 76 | 77 | 78 => {
                // --> 2190
                full_actions_2190(parser)
            }
            283 | 284 => {
                // --> 2191
                full_actions_2191(parser)
            }
            _ => panic!("Unexpected state value.")
        }
}
```

Without negative sign:
```rust
pub fn full_actions_2192<'alloc, Handler>(parser: &mut Handler) -> Result<'alloc, bool>
where
    Handler: ParserTrait<'alloc, StackValue<'alloc>> + AstBuilderDelegate<'alloc>
{
        match parser.top_state() {
            171 | 172 | 173 | 174 | 176 | 177 | 178 | 179 | 181 | 182 | 183 | 184 | 185 | 186 | 187 | 188 | 189 | 190 | 191 | 192 | 193 | 194 | 195 | 196 | 198 | 199 | 200 | 201 | 202 | 203 | 204 | 205 | 206 | 207 | 208 | 210 | 211 | 212 | 213 | 214 | 215 | 216 | 217 | 218 | 219 | 220 | 221 | 222 | 223 | 224 | 225 | 226 | 227 | 228 | 229 | 230 | 231 | 232 | 233 | 234 | 235 | 236 | 237 | 238 | 239 => {
                // --> 2187
                full_actions_2187(parser)
            }
            244 | 245 | 246 | 247 | 248 | 249 | 250 | 251 | 252 | 253 | 254 | 255 | 256 | 257 | 258 | 259 | 260 | 261 | 262 | 263 | 264 | 265 | 266 | 267 | 268 | 269 | 270 | 271 | 272 | 273 | 274 | 275 | 276 | 277 | 278 | 279 | 280 | 281 | 282 | 283 | 284 | 285 | 286 | 287 | 288 | 289 | 290 | 291 | 292 | 293 | 294 | 295 | 296 | 297 | 298 | 299 | 300 | 301 | 302 | 303 | 304 | 305 | 306 | 307 | 308 | 309 | 310 | 311 | 312 | 313 | 314 | 315 | 316 | 317 | 318 | 319 | 320 => {
                // --> 2188
                full_actions_2188(parser)
            }
            120 | 121 | 122 | 123 | 124 | 125 | 126 | 127 | 128 | 129 | 130 | 131 | 132 | 133 | 134 | 135 | 136 | 137 | 138 | 139 | 140 | 141 | 142 | 143 | 144 | 145 | 146 | 147 | 148 | 149 | 150 | 151 | 152 | 153 | 154 | 155 | 156 | 157 | 158 | 159 | 160 | 161 | 162 | 163 | 164 | 165 | 166 | 167 | 168 | 169 | 170 => {
                // --> 2189
                full_actions_2189(parser)
            }
            321 | 322 | 324 | 325 | 327 | 335 | 337 | 345 | 404 | 405 | 424 | 443 | 444 | 450 | 470 | 478 | 480 | 481 | 483 | 508 | 509 | 511 | 512 | 517 | 527 | 529 | 532 | 533 | 535 | 536 | 538 | 540 | 541 | 542 | 543 | 546 | 547 | 548 | 549 | 552 | 554 | 557 | 559 | 561 | 566 | 573 | 575 | 578 | 579 | 580 | 581 | 585 | 586 | 588 | 589 | 590 | 591 | 592 | 593 | 595 | 597 | 600 | 605 | 606 | 608 | 613 | 614 | 615 | 617 | 619 | 622 | 627 | 629 | 632 | 635 | 639 => {
                // --> 2190
                full_actions_2190(parser)
            }
            116 | 117 => {
                // --> 2191
                full_actions_2191(parser)
            }
            _ => panic!("Unexpected state value.")
        }
}
```